zkevm-circuits/src/pi_circuit2.rs

```diff
@@ -65,162 +33,114 @@ pub struct BlockValues {
+        use FieldType::*;
+        [
+            (
+                "l1_signal_service",
+                None,
+                self.l1_signal_service.to_be_bytes(),
+            ),
+            (
+                "l2_signal_service",
+                None,
+                self.l2_signal_service.to_be_bytes(),
+            ),
+            ("l2_contract", None, self.l2_contract.to_be_bytes()),
+            ("meta_hash", None, self.meta_hash.to_be_bytes()),
+            ("parent_hash", None, self.parent_hash.to_be_bytes()),
+            ("block_hash", BlockHash, self.block_hash.to_be_bytes()),
+            ("signal_root", None, self.signal_root.to_be_bytes()),
+            ("graffiti", None, self.graffiti.to_be_bytes()),
+            (
+                "prover+parentGasUsed+gasUsed",
+                None,
+                self.field9.to_be_bytes(),
+            ),
+        ]
     }
 
-    fn get_block_hash(
-        block: &witness::Block<F>,
-        prover: Address,
-        txs_hash: H256,
-    ) -> (Bytes, H256, F, F) {
-        let mut stream = RlpStream::new();
-        stream.begin_unbounded_list();
-        stream
-            .append(&block.eth_block.parent_hash)
-            .append(&*OMMERS_HASH);
-        rlp_opt(&mut stream, &block.eth_block.author);
-        stream
-            .append(&block.eth_block.state_root)
-            .append(&block.eth_block.transactions_root)
-            .append(&block.eth_block.receipts_root);
-        rlp_opt(&mut stream, &block.eth_block.logs_bloom);
-        stream.append(&block.eth_block.difficulty);
-        rlp_opt(&mut stream, &block.eth_block.number);
-        stream
-            .append(&block.eth_block.gas_limit)
-            .append(&block.eth_block.gas_used)
-            .append(&block.eth_block.timestamp)
-            .append(&block.eth_block.extra_data.as_ref());
-        rlp_opt(&mut stream, &block.eth_block.mix_hash);
-        rlp_opt(&mut stream, &block.eth_block.nonce);
-        // rlp_opt(&mut stream, &block.eth_block.base_fee_per_gas);
-        stream.append(&prover).append(&txs_hash);
-        stream.finalize_unbounded_list();
-        let out: bytes::Bytes = stream.out().into();
-        let rlp = out.into();
-        let hash = keccak256(&rlp);
-        // append prover and txs_hash
-        let (hi, lo) = Self::split_hash(hash);
-        (rlp, hash.into(), hi, lo)
+    fn rpi_bytes(&self) -> Vec<u8> {
+        iter::empty()
+            .chain(self.l1_signal_service.to_be_bytes())
+            .chain(self.l2_signal_service.to_be_bytes())
+            .chain(self.l2_contract.to_be_bytes())
+            .chain(self.meta_hash.to_be_bytes())
+            .chain(self.parent_hash.to_be_bytes())
+            .chain(self.block_hash.to_be_bytes())
+            .chain(self.signal_root.to_be_bytes())
+            .chain(self.graffiti.to_be_bytes())
+            .chain(self.field9.to_be_bytes())
+            .collect()
```

Comment:
Could just extract and collect these values from `assignments`, that way no need to keep both lists iterating over the same values in the same order consistent.